File: web/src/yngdieng/query-parser.ts

import { Query }  from "yngdieng/shared/services_pb";
import { getInitialFromString, getFinalFromString, getToneFromString } from './utils';

const ACCEPTED_KEYS = new Set(["i", "f", "t"]);

/*
 * Query format
 *
 * "我" => HanziQuery
 * 
 *"i:l f:uang t:up_falling" => Phonology Query
 * 
 * "nguai" => Fuzzy query
 */

/**
 * Parser for a query string.
 */
export class QueryParser {

    /**
     * Converts a query from string to Query.
     * 
     * 1) If it is key-value pairs and all keys are recognized.
     *    => PhonologyQuery
     * 2) If it is alphanumberic => Fuzzy Pinyin Query
     * 3) Otherwise => Hanzi Query
     * 
     * @returns null if the query string is invalid;
     */
    public parse(text: string): Query | null {
        let tryResult = this.tryParseAsKeyValuePairs(text);
        if (tryResult != null) {
            let phonologyQuery = new Query.PhonologyQuery();
            if (tryResult.has("i")) {
                let initial = getInitialFromString(tryResult.get("i"));
                if (initial === undefined) {
                    return null;
                }
                phonologyQuery.setInitial(initial); 
            }
            if (tryResult.has("f")) {
                let final = getFinalFromString(tryResult.get("f"))
                if (final === undefined) {
                    return null;
                }
                phonologyQuery.setFinal(final);
            }
            if (tryResult.has("t")) {
                let tone = getToneFromString(tryResult.get("t"));
                if (tone === undefined) {
                    return null;
                }
                phonologyQuery.setTone(tone);
            }
            let query = new Query();
            query.setPhonologyQuery(phonologyQuery);
            return query;
        }
        if (this.isAlphanumeric(text)) {
            let query = new Query();
            query.setFuzzyPronQuery(text.trim());
            return query;
        }
        let query = new Query();
        query.setHanziQuery(text.trim());
        return query;
    }

    private tryParseAsKeyValuePairs(text: string): Map<string, string> | null {
        let tokens = text
            .trim()
            .split(" ")
            .filter(t => t.length > 0);
        if (tokens.length == 0) {
            return null;
        }
        let output = new Map<string, string>();
        for (let token of tokens) {
            let columnPos = token.indexOf(":");
            if (columnPos < 0) {
                return null;
            }
            let key = token.substring(0, columnPos);
            let value = token.substring(columnPos + 1, token.length);
            if (!ACCEPTED_KEYS.has(key)) {
                return null;
            }
            output.set(key, value);
        }
        return output.size > 0 ? output : null;
    }

    private isAlphanumeric(s: string): boolean {
        return s.match(/^[a-z0-9\s]+$/i) != null;
    }
}

File: server/backend/Services/QueryParser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using Yngdieng.Protos;

namespace Yngdieng.Backend.Services
{
  public static class QueryParser
  {

    private static readonly Regex rAlphaNumeric = new Regex("^[a-zA-Z0-9]*$");

    /*
    * Query format
    *
    * "我" => HanziQuery
    *
    * "我 sort:phonology" => HanziQuery
    * 
    *  "i:l f:uang t:up_falling" => Phonology Query
    * 
    * "nguai" => Fuzzy query
    */

    /**
     * Converts a query from string to Query.
     * 
     * 1) If it is key-value pairs and all keys are recognized.
     *    => PhonologyQuery
     * 2) If it is alphanumberic => Fuzzy Pinyin Query
     * 3) Otherwise => Hanzi Query
     * 
     * @returns null if the query string is invalid;
     */
    public static Query? Parse(string text)
    {
      (var keyValuePairs, var remainingTokens) = ParseAsKeyValuePairs(text);

      var query = new Query();
      // Defaults
      query.SortBy = Query.Types.SortByMethod.InitialFinalTone;
      if (keyValuePairs.ContainsKey("sort"))
      {
        var sortMethod = keyValuePairs["sort"];
        var parsed = ParseSortByMethod(sortMethod);
        if (parsed != null)
        {
          query.SortBy = parsed ?? Query.Types.SortByMethod.SortByUnspecified;
        }
        else
        {
          return null;
        }
      }

      if (keyValuePairs.ContainsKey("historical"))
      {
          var value = keyValuePairs ["historical"]
                          .Trim()
                          .ToLowerInvariant();
          if (value == "yes" || value == "true")
          {
              query.AlwaysIncludeHistorical = true;
          }
          else if (value == "only")
          {
              query.AlwaysIncludeHistorical = true;
              query.OnlyHistorical = true;
          }
      }
      if (keyValuePairs.ContainsKey("i") || keyValuePairs.ContainsKey("f") || keyValuePairs.ContainsKey("t"))
      {
        var phonologyQuery = new Query.Types.PhonologyQuery();
        // Initial
        if (keyValuePairs.ContainsKey("i"))
        {
          var initial = PhonologyUtils.GetInitialFromString(keyValuePairs["i"]);
          if (initial == Initial.Unspecified)
          {
            return null;
          }
          phonologyQuery.Initial = initial;
        }
        // Final
        if (keyValuePairs.ContainsKey("f"))
        {
          var final = PhonologyUtils.GetFinalFromString(keyValuePairs["f"]);
          if (final == Final.Unspecified)
          {
            return null;
          }
          phonologyQuery.Final = final;
        }
        // Tone
        if (keyValuePairs.ContainsKey("t"))
        {
          var tone = PhonologyUtils.GetToneFromString(keyValuePairs["t"]);
          if (tone == Tone.Unspecified)
          {
            return null;
          }
          phonologyQuery.Tone = tone;
        }

        query.PhonologyQuery = phonologyQuery;
        return query;
      }

      if (remainingTokens.Length > 0 && remainingTokens.All(t => rAlphaNumeric.IsMatch(t)))
      {
        query.FuzzyPronQuery = String.Join(" ", remainingTokens);
        return query;
      }

      if (remainingTokens.Length == 1 && !rAlphaNumeric.IsMatch(remainingTokens[0]))
      {
        query.HanziQuery = remainingTokens[0];
        return query;
      }

      return null;
    }

    /// <summary>
    /// 
    /// </summary>
    /// <param name="text"></param>
    /// <returns>(key value pairs, remaining tokens)</returns>
    private static (IDictionary<string, string>, string[]) ParseAsKeyValuePairs(string text)
    {
      var tokens = text.Trim().Split(" ").Where(t => t.Length > 0);
      if (tokens.Count() == 0)
      {
        return (new Dictionary<string, string>(), new string[0]);
      }
      var keyValuePairs = new Dictionary<string, string>();
      var remainingTokens = new List<string>();
      foreach (var token in tokens)
      {
        var columnPos = token.IndexOf(":", StringComparison.Ordinal);
        if (columnPos < 0)
        {
          remainingTokens.Add(token);
          continue;
        }
        var key = token.Substring(0, columnPos);
        var value = token.Substring(columnPos + 1);
        keyValuePairs[key] = value;
      }

      return (keyValuePairs, remainingTokens.ToArray());
    }


    private static Query.Types.SortByMethod? ParseSortByMethod(string text)
    {
      switch (text.Trim().ToLower())
      {
        case "ift":
        case "phonology":
        case "initial_final_tone":
          return Query.Types.SortByMethod.InitialFinalTone;
        default:
          return null;
      }
    }

  }
}
